docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [Requirements (Optional)](#requirements-optional)
+      - [Functional Requirements](#functional-requirements)
+        - [FR1](#fr1)
+        - [FR2](#fr2)
+      - [Non-Functional Requirements](#non-functional-requirements)
+        - [NFR1](#nfr1)
+        - [NFR2](#nfr2)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+
+- **OID:** Object Identifier defined by the International Telecommunications Union and used in PKI to identify attributes on certificates.
+
+- **PKI:** Public Key Infrastructure
+
+- **TPM:** Trusted Platform Module (TPM) is a specification defined by the Trusted Computing Group (TCG) that allows hosts to attest to their identity via PKI and a secure crypto-processor which may either be a separate chip, built into the CPU or a virtual device provided by the hypervisor.
+
+
+## Summary
+
+This proposal outlines a method to secure node registration within Cluster API, to solve 3 primary problems:
+- Solving a class of attacks involving node impersonation allowing an attacker to access secrets and volumes they shouldn’t by using hardware attestation of node identity.
+- Prevent token reuse in MachinePools where the token cannot be easily rotated.
+
+
+## Motivation
+
+Cluster API default core components are largely reliant on kubeadm for cluster bootstrapping and node registration. Kubeadm is a platform-agnostic command line tool designed to assist users to bootstrap Kubernetes clusters, and is used as a building block in Cluster API.
+Because kubeadm is platform-independent and is intended to provide an “easy path” to cluster bootstrapping,  there are a number of inherent design decisions that limit the overall security of the provisioned cluster:
+- Kubeadm uses TLS bootstrapping for node registration, however the default workflow used by Cluster API uses bootstrap token which allow registration as arbitrary node names.
+    - When used in this mode, Kubeadm essentially does “client-side validation” to prevent node hijacking, but this does not mean the token cannot be reused by an attacker within the lifetime of the token to perform a hijack.
+    - Cluster API cannot scope a token down  to a specific node, because bootstrap providers, nor most infrastructure providers know the identity of the node ahead of time.
```

Comment:
```suggestion
    - Cluster API cannot scope a token down to a specific node, because bootstrap providers, nor most infrastructure providers know the identity of the node ahead of time.
```